stacker/hooks/aws_lambda.py

```diff
@@ -296,30 +338,27 @@ def _upload_function(s3_conn, bucket, prefix, name, options, follow_symlinks):
             through.
     """
     try:
-        root = os.path.expanduser(options['path'])
+        path = os.path.expanduser(options['path'])
     except KeyError as e:
         raise ValueError(
             "missing required property '{}' in function '{}'".format(
                 e.args[0], name))
 
-    includes = _check_pattern_list(options.get('include'), 'include',
-                                   default=['**'])
-    excludes = _check_pattern_list(options.get('exclude'), 'exclude',
-                                   default=[])
+    if not os.path.isabs(path):
+        path = os.path.abspath(os.path.join(get_config_directory(), path))
 
-    logger.debug('lambda: processing function %s', name)
+    if path.endswith('.zip') and os.path.isfile(path):
+        logging.debug('lambda: using prebuilt zip: %s', path)
 
-    # os.path.join will ignore other parameters if the right-most one is an
-    # absolute path, which is exactly what we want.
-    if not os.path.isabs(root):
-        root = os.path.abspath(os.path.join(get_config_directory(), root))
-    zip_contents, content_hash = _zip_from_file_patterns(root,
-                                                         includes,
-                                                         excludes,
-                                                         follow_symlinks)
+        return _upload_prebuilt_zip(s3_conn, bucket, prefix, name, options,
```

Comment:
Kinda bugs me that we have to methods that call `_upload_code`. I think `_upload_prebuilt_zip` and `_build_and_upload_zip` should be changed so that we return a tuple `(zip_file, version)` and then we call `_upload_code` once.

Reply:
I've kept them separated due to the prebuilt case passing the open file through instead of loading all the contents. When that happens `_upload_code` needs to be called in the scope of the context manager, which will require some weird conditionals if the call is unified.

Reply:
Gotcha. `_build_and_upload_zip` currently doesn't use the context manager when passing the zip file down to `_upload_code`. Think we should change it so we're consistent?